### PR TITLE
feat: Determine our `SwiftFormat` rules - WPB-10848

### PR DESCRIPTION
**Pull request by samwyndham**
# ⚠️ ⚠️ ⚠️ This PR will never be merged to `develop`. You don't need to review it closely. Please read the description carefully.

### Issue

As discussed in [Linting & Formatting Improvements](https://wearezeta.atlassian.net/wiki/spaces/IOS/pages/1391755277/Linting+Formatting+Improvements) (@johnxnguyen please take a look at that document as you were away during the meeting) we plan to improve our linting and formatting of Swift code to cover the entire codebase with more rules.

This PR is a place of discussion of `SwiftFormat` rules only. A similar PR will be created in the future to discuss `SwiftLint` rules. Once we have consensus on both `SwiftFormat` and `SwiftLint` rules we can move forward. 

Nearly every rule is enabled in this PR including the ones the make me 🤮 - there are two exceptions:

- [acronyms](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#acronyms) as this is hard to use with coding keys.
- [docComments](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#doccomments) as this won't work with our _floating_ style comments.

Personally I'm of the opinion that generally the more rules the better that lead to a standard style across the team, but if we have preferences which these rules go against perhaps we can assert them via linting instead.

#### Task

Each commit (approximately) in this PR introduces a new rule to the codebase. The rule name will be in the title of the commit and there may be some additional info about it in the commit description.

As a reviewer your job is to go through the PR **commit by commit** and note down any rules that you dislike or have concerns about. 

⚠️ Please leave you review here: https://wearezeta.atlassian.net/wiki/spaces/IOS/pages/1454800899/SwiftFormat+review

I will leave this PR up for some time. Please enjoy bike shedding.

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

**Comment by netbe** [comment]
> [docComments](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#doccomments) as this won't work with our floating style comments.
what do you mean by that?

File: wire-ios-request-strategy/Sources/Helpers/ZMSound.swift [hunk 1]
```diff
@@ -76,7 +76,9 @@ public enum ZMSound: String, CustomStringConvertible {
         switch self {
         case .None:
             return nil
-        case .WireCall, .WirePing, .WireText:
+        case .WireCall,
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Helpers/ZMSound.swift:79]: I'll go with like or not enforces, worst case a line length would enforce going to the next line

File: wire-ios-cryptobox/WireCryptobox/EncryptionSessionsDirectory.swift [hunk 2]
```diff
@@ -31,14 +31,15 @@ enum EncryptionSessionError: Int {
...
-        }
+        var info =
+            switch self {
+            case .unknown:
+                "Unknown EncryptionSessionError"
+            case .encryptionFailed:
+                "Encryption Failed"
+            case .decryptionFailed:
+                "Decryption Failed"
+            }
```
Comment (netbe) [inline wire-ios-cryptobox/WireCryptobox/EncryptionSessionsDirectory.swift:42]: don't like this

**Comment by samwyndham** [comment]
> > [docComments](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#doccomments) as this won't work with our floating style comments.
> > what do you mean by that?

@netbe I mean that we often have a line of whitespace between the doc comment and the declaration:

```swift
/// Adds two numbers

func add(a: Int, b: Int) {
  // ...
}
```

Whereas the rule expects the following and behaves incorrectly without it.

```swift
/// Adds two numbers
func add(a: Int, b: Int) {
  // ...
}

**Review by samwyndham — COMMENTED** [review]
Here is my review of the rules. I am generally happy with most rules. I have reservations about the following:

- [conditionalAssignment](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#conditionalAssignment) - I have mixed feelings about this one. I'm not super against it but feel that in some cases the content can become overly dense.
- [preferForLoop](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#preferForLoop) - Mixed feelings. I would prefer if it ignored one liner `forEach` statements which can be done with `--onelineforeach` option.
- [redundantType](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#redundantType) - This can be bad for compile times. If we do enable it we should also measure our compile times and add exceptions where necessary.
- [markTypes](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#markTypes) - Creates a lot of marks and I don't feel they are that useful for anything. If everything becomes a mark what is the point.
- [organizeDeclarations](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#organizeDeclarations) - Not against this, but I think this one requires a closer look and discussion. It moves our code around a lot.
- [wrapConditionalBodies](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#wrapConditionalBodies) - I don't like this one as I like single line early exits.
- [wrapSwitchCases](https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#wrapSwitchCases) - I don't like this one. It takes up unnecessary space in my opinion.

File: wire-ios-data-model/Source/MLS/MLSGroup.swift [hunk 3]
```diff
@@ -60,11 +60,9 @@ public class MLSGroup: ZMManagedObject {
 
     static let idKey = "id"
 
-    @NSManaged
-    private var primitiveId: Data?
+    @NSManaged private var primitiveId: Data?
```
Comment (netbe) [inline wire-ios-data-model/Source/MLS/MLSGroup.swift]: weird it does not behave as @discardableResult or @objc

**Review by netbe — COMMENTED** [review]
Reviewed all the rules

**Review by KaterinaWire — APPROVED** [review]

File: .swiftformat [hunk 4]
```diff
@@ -20,7 +20,8 @@
     enumNamespaces, \
     extensionAccessControl, \
     genericExtensions, \
-    hoistAwait
+    hoistAwait, \
+    hoistPatternLet
```
Comment (caldrian) [inline .swiftformat]: This rule I would probably disable and decide on a case-by-case basis.

File: WireUI/Sources/WireReusableUIComponents/AccessoryTextField/AccessoryTextField.swift [hunk 5]
```diff
@@ -154,12 +154,12 @@ extension AccessoryTextField {
 // MARK: - Custom edge insets
 
 extension AccessoryTextField {
-    public override func textRect(forBounds bounds: CGRect) -> CGRect {
+    override public func textRect(forBounds bounds: CGRect) -> CGRect {
```
Comment (caldrian) [inline WireUI/Sources/WireReusableUIComponents/AccessoryTextField/AccessoryTextField.swift:173]: I suggest to configure the `modifierOrder` rule with a different order. At least `override` I would want to move to the right.

File: .swiftformat [hunk 6]
```diff
@@ -28,7 +28,8 @@
     leadingDelimiters, \
     linebreakAtEndOfFile, \
     linebreaks, \
-    modifierOrder
+    modifierOrder, \
+    numberFormatting
```
Comment (caldrian) [inline .swiftformat]: I like the rule, however, we should discuss about the configuration.

File: .swiftformat [hunk 7]
```diff
@@ -87,6 +87,7 @@
     # OPTIONAL
     # acronyms, \ - Difficult to use because it messes with coding keys etc. So a lot of disabling this rule would be necessary
     blankLineAfterSwitchCase, \
+    blankLinesBetweenImports, \
```
Comment (caldrian) [inline .swiftformat:90]: I would disable this rule. I really like to have `import` and `@testable import` separated and sorted within their blocks.

File: .swiftformat [hunk 8]
```diff
@@ -91,6 +91,7 @@
     blockComments, \
     # docComments - Cannot enable due to our floating doc comment style
     isEmpty, \ # Rare false positives in situations where a type provides a `count` but not a `isEmpty` property.
+    markTypes, \
```
Comment (caldrian) [inline .swiftformat:94]: I don't consider this rule beneficial. Xcode would then show the name of the type twice, right? 
Screenshot 2024-10-21 at 7 22 42 PM

File: .swiftformat [hunk 9]
```diff
@@ -96,6 +96,7 @@
     organizeDeclarations, \
     redundantProperty, \
     sortSwitchCases, \
+    wrapConditionalBodies, \
```
Comment (caldrian) [inline .swiftformat:99]: I wouldn't enable this rule.

File: .swiftformat [hunk 10]
```diff
@@ -97,6 +97,7 @@
     redundantProperty, \
     sortSwitchCases, \
     wrapConditionalBodies, \
+    wrapEnumCases, \ # In rare cases this messes up enums that fail the rule and has to be manually fixed - affected 4 enums over the entire project
```
Comment (caldrian) [inline .swiftformat:100]: I wouldn't enable this rule.

File: .swiftformat [hunk 11]
```diff
@@ -99,6 +99,7 @@
     wrapConditionalBodies, \
     wrapEnumCases, \ # In rare cases this messes up enums that fail the rule and has to be manually fixed - affected 4 enums over the entire project
     wrapMultilineConditionalAssignment, \
+    wrapSwitchCases, \
```
Comment (caldrian) [inline .swiftformat:102]: I probably wouldn't enable this rule.

**Review by caldrian — APPROVED** [review]
Great work! Looking forward to enabling many of these rules.

File: .swiftformat [hunk 12]
```diff
@@ -14,11 +14,14 @@
...
+    elseOnSameLine
 
 # Options
 
 --lineaftermarks true # blankLinesAroundMark
 --typeblanklines remove # blankLinesAtStartOfScope
 --allman false # braces
 --condassignment after-property # conditionalAssignment
+--elseposition same-line # elseOnSameLine
+--guardelse auto # elseOnSameLine
```
Comment (jullianm) [inline .swiftformat:111]: 👍 I'm surprised there's not many fixes in codebase after adding this rule

File: WireAPI/Tests/WireAPITests/APIs/TeamsAPI/TeamsAPITests.swift [hunk 13]
```diff
@@ -55,7 +55,7 @@ final class TeamsAPITests: XCTestCase {
 
     func testGetTeamMembersRequest() async throws {
         try await apiSnapshotHelper.verifyRequestForAllAPIVersions { sut in
-            _ = try await sut.getTeamMembers(for: .mockID1, maxResults: 2_000)
+            _ = try await sut.getTeamMembers(for: .mockID1, maxResults: 2000)
```
Comment (jullianm) [inline WireAPI/Tests/WireAPITests/APIs/TeamsAPI/TeamsAPITests.swift:57]: there's another rule that does the exact opposite right ? like formatting 2000 into 2_000, I think it makes sense for big numbers, 1_000_000 is more readable but for small numbers 1000 is fine

File: wire-ios-data-model/Source/MLS/CoreCryptoKeyProvider.swift [hunk 14]
```diff
@@ -87,7 +87,7 @@ struct CoreCryptoKeychainItem: KeychainItemProtocol {
         ]
     }
 
-    func setQuery<T>(value: T) -> [CFString: Any] {
+    func setQuery(value: some Any) -> [CFString: Any] {
```
Comment (jullianm) [inline wire-ios-data-model/Source/MLS/CoreCryptoKeyProvider.swift:100]: it doesn't make sense, I think it should resolve this to `Any`, not `some Any`

File: .swiftformat [hunk 15]
```diff
@@ -30,7 +30,8 @@
     linebreaks, \
     modifierOrder, \
     numberFormatting, \
-    opaqueGenericParameters
+    opaqueGenericParameters, \
+    preferForLoop
```
Comment (jullianm) [inline .swiftformat]: not particularly fan of this rule, in many cases it makes sense to use ForEach() over a for-in loop.

File: wire-ios-data-model/Source/Model/Conversation/ConversationLike.swift [hunk 16]
```diff
@@ -88,7 +88,7 @@ extension ZMConversation: ConversationLike {
 
     public var sortedServiceUsers: [UserType] {
         localParticipants
-            .filter { $0.isServiceUser }
+            .filter(\.isServiceUser)
```
Comment (jullianm) [inline wire-ios-data-model/Source/Model/Conversation/ConversationLike.swift:100]: 👍

File: .swiftformat [hunk 17]
```diff
@@ -50,3 +50,4 @@
 --extensionacl on-declarations # extensionAccessControl
 --patternlet hoist # hoistPatternLet
 --indent 4 # indent
+--anonymousforeach ignore # preferForLoop
```
Comment (jullianm) [inline .swiftformat:117]: not particularly fan of enforcing the usage of one or the other, II would not add any of these two rules regarding loops.

File: .swiftformat [hunk 18]
```diff
@@ -67,3 +68,4 @@
 --indent 4 # indent
 --anonymousforeach ignore # preferForLoop
 # --self init-only # redundantSelf TODO
+--redundanttype inferred
```
Comment (jullianm) [inline .swiftformat]: hmmm interesting, I feel like in some cases we might want the type to be explicit

```
var number = 2 // inferred as Int
var number: Double = 2
```

or 

```
var id: UUID!
let copy = id // inferred as UUID? but we know this should have a value
let copy: UUID = id // not having to deal with optional unwrapping anymore
```

File: .swiftformat [hunk 19]
```diff
@@ -69,7 +69,8 @@
     spaceInsideComments, \
     spaceInsideGenerics, \
     spaceInsideParens, \
-    strongOutlets
+    strongOutlets, \
```
Comment (jullianm) [inline .swiftformat:72]: 🤔

File: .swiftformat [hunk 20]
```diff
@@ -74,7 +74,8 @@
     todos, \
     trailingClosures, \
     trailingCommas, \
-    trailingSpace
+    trailingSpace, \
+    typeSugar
```
Comment (jullianm) [inline .swiftformat]: ❤️

File: .swiftformat [hunk 21]
```diff
@@ -77,8 +77,9 @@
     trailingSpace, \
     typeSugar, \
     unusedArguments, \
-    void
-
+    void, \
+    wrap, \
+    wrapArguments
```
Comment (jullianm) [inline .swiftformat]: definitely yes on this rule

File: .swiftformat [hunk 22]
```diff
@@ -82,7 +82,8 @@
     wrapArguments, \
     wrapAttributes, \
     wrapLoopBodies, \
-    wrapSingleLineComments
+    wrapSingleLineComments, \
+    yodaConditions
```
Comment (jullianm) [inline .swiftformat]: let's add this rule just for its name.

File: .swiftformat [hunk 23]
```diff
@@ -85,6 +85,7 @@
     wrapSingleLineComments, \
     yodaConditions, \
     # OPTIONAL
+    # acronyms, \ - Difficult to use because it messes with coding keys etc. So a lot of disabling this rule would be necessary
```
Comment (jullianm) [inline .swiftformat]: so maybe let's not add this rule.

File: .swiftformat [hunk 24]
```diff
@@ -93,6 +93,7 @@
     isEmpty, \ # Rare false positives in situations where a type provides a `count` but not a `isEmpty` property.
     markTypes, \
     noExplicitOwnership, \
+    organizeDeclarations, \
```
Comment (jullianm) [inline .swiftformat:96]: I find this rule useful, we often see discrepancies in codebase regarding the way we write our // MARK.

File: .swiftformat [hunk 25]
```diff
@@ -98,6 +98,7 @@
     sortSwitchCases, \
     wrapConditionalBodies, \
     wrapEnumCases, \ # In rare cases this messes up enums that fail the rule and has to be manually fixed - affected 4 enums over the entire project
+    wrapMultilineConditionalAssignment, \
```
Comment (jullianm) [inline .swiftformat:101]: Seeing the changes below, not a big fan of this rule.

**Review by jullianm — APPROVED** [review]
Nice work 👍

**Review by johnxnguyen — COMMENTED** [review]
Left my review in confluence

File: WireAPI/Tests/WireAPITests/APIs/TeamsAPI/TeamsAPITests.swift (hunk 13 above)
Comment (samwyndham) [reply WireAPI/Tests/WireAPITests/APIs/TeamsAPI/TeamsAPITests.swift:57]: This is configurable. I'll put you down as in favor but with a different configuration.

File: wire-ios-data-model/Source/MLS/CoreCryptoKeyProvider.swift (hunk 14 above)
Comment (samwyndham) [reply wire-ios-data-model/Source/MLS/CoreCryptoKeyProvider.swift:100]: I think that is the new generics API if we don't constrain the generic to a specific type.

File: .swiftformat (hunk 19 above)
Comment (samwyndham) [reply .swiftformat:72]: There was a bit of conversation about this. Apparently this is what apple advises in most cases now. However in practice we don't have any weak IBOutlets in our codebase and probably won't get any.

File: wire-ios-data-model/Source/ManagedObjectContext/Migration/CoreDataMessagingMigrationVersion.swift [hunk 26]
```diff
@@ -85,8 +85,8 @@ enum CoreDataMessagingMigrationVersion: String, CoreDataMigrationVersion {
             nil
         case .v116, .v117:
             .v118
-        case .v115,
-             .v114:
+        case .v114,
```
Comment (netbe) [inline wire-ios-data-model/Source/ManagedObjectContext/Migration/CoreDataMessagingMigrationVersion.swift:89]: here the order is changed

**Pull request closed without merging**